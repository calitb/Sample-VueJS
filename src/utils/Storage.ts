export function removeItem(key: string) {
  sessionStorage.removeItem(key);
}

export function getItem(key: string): string | null {
  return sessionStorage.getItem(key);
}

export function setItem(key: string, value: string) {
  sessionStorage.setItem(key, value);
}
