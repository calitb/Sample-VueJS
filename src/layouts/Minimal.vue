<template>
  <div>
    <b-navbar toggleable="lg" type="dark" variant="dark">
      <router-link :to="{ name: 'list' }">
        <b-navbar-brand>VueJS Tutorial</b-navbar-brand>
      </router-link>
      <b-navbar-toggle v-if="authenticated()" target="nav-collapse"></b-navbar-toggle>
      <b-collapse v-if="authenticated()" id="nav-collapse" is-nav>
        <b-navbar-nav class="ml-auto">
          <b-button variant="outline-light" @click="logout">Sign Out</b-button>
        </b-navbar-nav>
      </b-collapse>
    </b-navbar>
    <router-view />
  </div>
</template>

<script lang="ts">
import Vue from 'vue';

import { getItem, removeItem } from '@/utils/Storage';

export default Vue.extend({
  name: 'app-layout-minimal',
  methods: {
    logout(): void {
      removeItem('SESSION');
      this.$router.push({ name: 'login' });
    },
    authenticated(): boolean {
      return getItem('SESSION') !== null;
    }
  }
});
</script>
