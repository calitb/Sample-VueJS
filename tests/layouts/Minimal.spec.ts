import { shallowMount, mount, Wrapper, RouterLinkStub } from '@vue/test-utils';
import Minimal from '@/layouts/Minimal.vue';

import { createLocalVue } from '../utils';
import VueRouter from 'vue-router';

import * as Storage from '@/utils/Storage';

describe('Minimal Layout', () => {
  describe('Not logged', () => {
    let spyGetItem: jest.SpyInstance<string | null, [string]>;
    let wrapper: Wrapper<any>;

    const localVue = createLocalVue();
    localVue.use(VueRouter);

    beforeAll(() => {
      spyGetItem = jest.spyOn(Storage, 'getItem').mockImplementation((key: string) => {
        if (key === 'SESSION') {
          return null;
        }
        return 'error';
      });
      wrapper = shallowMount(Minimal, { localVue });
    });

    afterAll(() => {
      spyGetItem.mockRestore();
    });

    it('should get the session status', () => {
      expect(spyGetItem).toBeCalledTimes(2);
      expect(spyGetItem).toHaveBeenNthCalledWith(1, 'SESSION');
      expect(spyGetItem).toHaveBeenNthCalledWith(2, 'SESSION');
    });

    it('should render the component', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('should not have a logout button', () => {
      const button = wrapper.find('b-button-stub');
      expect(button.exists()).toBeFalsy();
    });
  });

  describe('Logged', () => {
    let spyGetItem: jest.SpyInstance<string | null, [string]>;
    let spyRemoveItem: jest.SpyInstance<void, [string]>;
    let wrapper: Wrapper<any>;

    const localVue = createLocalVue();

    const pushHandler = jest.fn();
    const $router = {
      push: pushHandler
    };

    beforeAll(() => {
      spyGetItem = jest.spyOn(Storage, 'getItem').mockImplementation((key: string) => {
        if (key === 'SESSION') {
          return 'true';
        }
        return null;
      });
      spyRemoveItem = jest.spyOn(Storage, 'removeItem').mockImplementation((key: string) => {});

      wrapper = mount(Minimal, {
        localVue,
        stubs: {
          'router-view': true,
          RouterLink: RouterLinkStub
        },
        mocks: {
          $router
        }
      });
    });

    afterAll(() => {
      spyGetItem.mockRestore();
      spyRemoveItem.mockRestore();
    });

    it('should get the session status', () => {
      expect(spyGetItem).toBeCalledTimes(2);
      expect(spyGetItem).toHaveBeenNthCalledWith(1, 'SESSION');
      expect(spyGetItem).toHaveBeenNthCalledWith(2, 'SESSION');
    });

    it('should render the component', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('should perform the logout when button is clicked', () => {
      const buttons = wrapper.findAll('button');
      expect(buttons).toHaveLength(2);

      buttons.at(1).trigger('click');

      expect(spyRemoveItem).toBeCalledTimes(1);
      expect(spyRemoveItem).toBeCalledWith('SESSION');

      expect(pushHandler).toBeCalledTimes(1);
      expect(pushHandler).toBeCalledWith({ name: 'login' });
    });

    it('should render the link to home', () => {
      expect(wrapper.findComponent(RouterLinkStub).props().to).toEqual({ name: 'list' });
    });
  });
});
