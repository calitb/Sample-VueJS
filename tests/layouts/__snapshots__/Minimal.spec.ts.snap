// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Minimal Layout Logged should render the component 1`] = `
<div>
  <nav
    class="navbar navbar-dark bg-dark navbar-expand-lg"
  >
    <a>
      <div
        class="navbar-brand"
      >
        VueJS Tutorial
      </div>
    </a>
     
    <button
      aria-controls="nav-collapse"
      aria-expanded="false"
      aria-label="Toggle navigation"
      class="navbar-toggler collapsed"
      type="button"
    >
      <span
        class="navbar-toggler-icon"
      />
    </button>
     
    <transition-stub
      css="true"
      enteractiveclass="collapsing"
      enterclass=""
      entertoclass="collapse show"
      leaveactiveclass="collapsing"
      leaveclass="collapse show"
      leavetoclass="collapse"
    >
      <div
        class="navbar-collapse collapse"
        id="nav-collapse"
        style="display: none;"
      >
        <ul
          class="navbar-nav ml-auto"
        >
          <button
            class="btn btn-outline-light"
            type="button"
          >
            Sign Out
          </button>
        </ul>
      </div>
    </transition-stub>
  </nav>
   
  <router-view-stub />
</div>
`;

exports[`Minimal Layout Not logged should render the component 1`] = `
<div>
  <b-navbar-stub
    tag="nav"
    toggleable="lg"
    type="dark"
    variant="dark"
  >
    <router-link-stub
      ariacurrentvalue="page"
      event="click"
      tag="a"
      to="[object Object]"
    >
      <b-navbar-brand-stub
        tag="div"
        target="_self"
      >
        VueJS Tutorial
      </b-navbar-brand-stub>
    </router-link-stub>
     
    <!---->
     
    <!---->
  </b-navbar-stub>
   
  <router-view-stub
    name="default"
  />
</div>
`;
